// @datadog/browser-logs/cjs/domain/assembly.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startLogsAssembly = startLogsAssembly;
const browser_core_1 = require("@datadog/browser-core");
const logger_1 = require("./logger");
const rumInternalContext_1 = require("./contexts/rumInternalContext");
function startLogsAssembly(sessionManager, configuration, lifeCycle, getCommonContext, reportError) {
    const statusWithCustom = logger_1.STATUSES.concat(['custom']);
    const logRateLimiters = {};
    statusWithCustom.forEach((status) => {
        logRateLimiters[status] = (0, browser_core_1.createEventRateLimiter)(status, configuration.eventRateLimiterThreshold, reportError);
    });
    lifeCycle.subscribe(0 /* LifeCycleEventType.RAW_LOG_COLLECTED */, ({ rawLogsEvent, messageContext = undefined, savedCommonContext = undefined, domainContext }) => {
        var _a, _b;
        const startTime = (0, browser_core_1.getRelativeTime)(rawLogsEvent.date);
        const session = sessionManager.findTrackedSession(startTime);
        const shouldSendLog = sessionManager.findTrackedSession(startTime, { returnInactive: true });
        if (!shouldSendLog) {
            return;
        }
        const commonContext = savedCommonContext || getCommonContext();
        let account;
        if (!(0, browser_core_1.isEmptyObject)(commonContext.account) && commonContext.account.id) {
            account = commonContext.account;
        }
        if (session && session.anonymousId && !commonContext.user.anonymous_id) {
            commonContext.user.anonymous_id = session.anonymousId;
        }
        const log = (0, browser_core_1.combine)({
            service: configuration.service,
            session_id: session ? session.id : undefined,
            session: session ? { id: session.id } : undefined,
            // Insert user and account first to allow overrides from global context
            usr: !(0, browser_core_1.isEmptyObject)(commonContext.user) ? commonContext.user : undefined,
            account,
            view: commonContext.view,
        }, commonContext.context, (0, rumInternalContext_1.getRUMInternalContext)(startTime), rawLogsEvent, messageContext);
        if (((_a = configuration.beforeSend) === null || _a === void 0 ? void 0 : _a.call(configuration, log, domainContext)) === false ||
            (log.origin !== browser_core_1.ErrorSource.AGENT &&
                ((_b = logRateLimiters[log.status]) !== null && _b !== void 0 ? _b : logRateLimiters['custom']).isLimitReached())) {
            return;
        }
        lifeCycle.notify(1 /* LifeCycleEventType.LOG_COLLECTED */, log);
    });
}
