// @datadog/browser-logs/cjs/domain/logsTelemetry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startLogsTelemetry = startLogsTelemetry;
const browser_core_1 = require("@datadog/browser-core");
const rumInternalContext_1 = require("./contexts/rumInternalContext");
const configuration_1 = require("./configuration");
function startLogsTelemetry(initConfiguration, configuration, reportError, pageExitObservable, session) {
    const telemetry = (0, browser_core_1.startTelemetry)("browser-logs-sdk" /* TelemetryService.LOGS */, configuration);
    telemetry.setContextProvider(() => {
        var _a, _b, _c, _d, _e, _f;
        return ({
            application: {
                id: (_a = (0, rumInternalContext_1.getRUMInternalContext)()) === null || _a === void 0 ? void 0 : _a.application_id,
            },
            session: {
                id: (_b = session.findTrackedSession()) === null || _b === void 0 ? void 0 : _b.id,
            },
            view: {
                id: (_d = (_c = (0, rumInternalContext_1.getRUMInternalContext)()) === null || _c === void 0 ? void 0 : _c.view) === null || _d === void 0 ? void 0 : _d.id,
            },
            action: {
                id: (_f = (_e = (0, rumInternalContext_1.getRUMInternalContext)()) === null || _e === void 0 ? void 0 : _e.user_action) === null || _f === void 0 ? void 0 : _f.id,
            },
        });
    });
    const cleanupTasks = [];
    if ((0, browser_core_1.canUseEventBridge)()) {
        const bridge = (0, browser_core_1.getEventBridge)();
        const telemetrySubscription = telemetry.observable.subscribe((event) => bridge.send('internal_telemetry', event));
        cleanupTasks.push(() => telemetrySubscription.unsubscribe());
    }
    else {
        const telemetryBatch = (0, browser_core_1.startBatchWithReplica)(configuration, {
            endpoint: configuration.rumEndpointBuilder,
            encoder: (0, browser_core_1.createIdentityEncoder)(),
        }, configuration.replica && {
            endpoint: configuration.replica.rumEndpointBuilder,
            encoder: (0, browser_core_1.createIdentityEncoder)(),
        }, reportError, pageExitObservable, session.expireObservable);
        cleanupTasks.push(() => telemetryBatch.stop());
        const telemetrySubscription = telemetry.observable.subscribe((event) => telemetryBatch.add(event, (0, browser_core_1.isTelemetryReplicationAllowed)(configuration)));
        cleanupTasks.push(() => telemetrySubscription.unsubscribe());
    }
    (0, browser_core_1.drainPreStartTelemetry)();
    (0, browser_core_1.addTelemetryConfiguration)((0, configuration_1.serializeLogsConfiguration)(initConfiguration));
    return {
        telemetry,
        stop: () => {
            cleanupTasks.forEach((task) => task());
        },
    };
}
